compiler/z/codegen/OMRPeephole.cpp

```diff
@@ -1861,3 +1871,133 @@ OMR::Z::Peephole::tryToRemoveRedundantLTR()
+               break;
+            case TR::InstOpCode::LH:
+               if (isSigned)
+                  {
+                  mr = current->getMemoryReference();
+                  mr->resetMemRefUsedBefore();
+                  replacement = generateRXInstruction(self()->cg(), TR::InstOpCode::LGH, current->getNode(), lgfrReg, mr, current->getPrev());
+                  }
+               break;
+            case TR::InstOpCode::LLH:
+               if (!isSigned)
+                  {
+                  mr = current->getMemoryReference();
+                  mr->resetMemRefUsedBefore();
+                  replacement = generateRXInstruction(self()->cg(), TR::InstOpCode::LLGH, current->getNode(), lgfrReg, mr, current->getPrev());
+                  }
+               break;
+            case TR::InstOpCode::LB:
+               if (isSigned)
+                  {
+                  mr = current->getMemoryReference();
+                  mr->resetMemRefUsedBefore();
+                  replacement = generateRXInstruction(self()->cg(), TR::InstOpCode::LGB, current->getNode(), lgfrReg, mr, current->getPrev());
+                  }
+               break;
+            case TR::InstOpCode::LLC:
+               if (isSigned)
+                  {
+                  mr = current->getMemoryReference();
+                  mr->resetMemRefUsedBefore();
+                  replacement = generateRXInstruction(self()->cg(), TR::InstOpCode::LLGC, current->getNode(), lgfrReg, mr, current->getPrev());
+                  }
+               break;
+            case TR::InstOpCode::XR:
+               if (lgfrReg == current->getRegisterOperand(2))
+                  replacement = generateRRInstruction(self()->cg(), TR::InstOpCode::XGR, current->getNode(), lgfrReg, lgfrReg, current->getPrev());
+               break;
+            case TR::InstOpCode::IILF:
+               replacement = generateRILInstruction(self()->cg(), isSigned ? TR::InstOpCode::LGFI : TR::InstOpCode::LLILF, current->getNode(), lgfrReg, toS390RILInstruction(current)->getSourceImmediate(), current->getPrev());
+               break;
+            case TR::InstOpCode::LHI:
+               if (isSigned)
+                  {
+                  replacement = generateRIInstruction(self()->cg(), TR::InstOpCode::LGHI, current->getNode(), lgfrReg, toS390RIInstruction(current)->getSourceImmediate(), current->getPrev());
+                  }
+               else
+                  {
+                  // The following sequence of instructions:
+                  //    LHI      GPR1, IMM   ; sign extend IMM from 16 to 32 bits
+                  //    LLGFR    GPR1, GPR1  ; zero extend from 32 to 64 bits
+                  // Can be converted to
+                  //    LLILF    GPR1, IMM'  ; where IMM' is IMM sign extended from 16 to 32 bits
+                  int16_t imm = toS390RIInstruction(current)->getSourceImmediate();
```

Comment:
The 32 bit immediate stored in the `TR::S390RIInstruction` object for `LHI` is not necessarily a correctly sign extended 16 bit value. For example, the object generated at [1] with have an immediate value of 0x0000FFFF, but the `LHI` instruction will load the value 0xFFFFFFFF into the target register. To avoid this problem I ensure the 16 bit value is sign extended correctly here.

[1] https://github.com/eclipse-openj9/openj9/blob/27dac39d2c5b8ea2df7c7e005a5ac0d9eb1f5b9c/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L1797